lib/services/saml_test.go

```diff
@@ -66,3 +69,69 @@ func TestCheckSAMLEntityDescriptor(t *testing.T) {
 	require.NoError(t, err)
 	require.Len(t, certs, 1)
 }
+
+func TestValidateRoles(t *testing.T) {
+	t.Parallel()
+
+	// Create a roleSet with <nil> role values as ValidateSAMLRole does
+	// not care what the role value is, just that it exists and that
+	// the RoleGetter does not return an error.
+	var validRoles roleSet = map[string]types.Role{
+		"foo": nil,
+		"bar": nil,
+	}
+
+	testCases := []struct {
+		desc      string
+		roles     []string
+		expectErr bool
```

Comment:
:nail_care: 
We could change this to be an `require.ErrorAssertionFunc` :

```suggestion
		errorCheck require.ErrorAssertionFunc
```
Each test would than need a `require.NoError` or 
```go
errorCheck: func(t require.TestingT, err error, i ...interface{}) {
				require.Error(t, err)
				require.ErrorIs(t, <some error>, err)
			},
```

And then, when checking for the error:
```go
tt.errorCheck(t, err)
```

Reply:
I'm on the fence about this - I prefer tests to be simple and obvious and I find that using higher-order functions here obscures the test. It's a pretty simple and obvious check as-is and I"m not sure how using the higher-order function here helps. It also means the zero value for the errorCheck function is not useful and it is a value that needs to be supplied to all test cases for the common case.

However I am not actually checking the error type in my test, so I've simplified the tests a little by just using `require.ErrorIs` which works when comparing nil as well as specific errors, removing the conditional and using the zero value effectively.

Reply:
Looks good :+1: